maven_iuvs/files.py

```diff
@@ -0,0 +1,444 @@
+
+        Properties
+        ----------
+        absolute_file_paths: list
+            Absolute file paths of all files matching patterns in path.
+        filenames: list
+            Filenames of all files matching patterns in path.
+
+        Notes
+        -----
+        This class used glob-style matching, so a pattern of ['**/*.pdf'] will recursively search for .pdf files
+        starting from path.
+        """
+        self.__check_path_exists(path)
+        self.__absolute_paths = self.__get_absolute_file_paths_matching_patterns_in_path(path, patterns)
+        self.__filenames = self.__get_filenames_from_absolute_paths(self.__absolute_paths)
+        self.__raise_value_error_if_no_files_found(self.__absolute_paths)
+
+    @property
+    def absolute_file_paths(self):
+        """ Get absolute_file_paths """
+        return self.__absolute_paths
+
+    @property
+    def filenames(self):
+        """ Get filenames """
+        return self.__filenames
+
+    def get_absolute_paths_of_filenames_containing_pattern(self, pattern):
+        """ Get the absolute paths of filenames that contain a requested pattern.
+
+        Parameters
+        ----------
+        pattern: str
+            A regex pattern to search filenames for.
+
+        Returns
+        -------
+        matching_file_paths: list
+            A list of absolute paths of filenames that match pattern.
+        """
+        try:
+            matching_file_paths = []
+            for counter, file in enumerate(self.__filenames):
+                if fnm.fnmatch(file, pattern):
+                    matching_file_paths.append(self.__absolute_paths[counter])
+            self.__warn_if_no_files_found(matching_file_paths)
+            return sorted(matching_file_paths)
+        except TypeError:
+            raise TypeError('pattern must be a string.')
+
+    def get_filenames_containing_pattern(self, pattern):
+        """ Get the filenames that contain a requested pattern.
+
+        Parameters
+        ----------
+        pattern: str
+            A regex pattern to search filenames for.
+
+        Returns
+        -------
+        matching_filenames: list
+            A list of filenames that match pattern.
+        """
+        absolute_paths = self.get_absolute_paths_of_filenames_containing_pattern(pattern)
+        matching_filenames = self.__get_filenames_from_absolute_paths(absolute_paths)
+        return matching_filenames
+
+    @staticmethod
+    def __check_path_exists(path):
+        try:
+            if not os.path.exists(path):
+                raise OSError(f'The path {path} does not exist on this computer.')
+        except TypeError:
+            raise TypeError('The input value of path should be a string.')
+
+    @staticmethod
+    def __get_absolute_file_paths_matching_patterns_in_path(path, patterns):
+        try:
+            all_paths = []
+            for i in patterns:
+                all_paths.extend(glob.glob(os.path.join(path, i), recursive=True))
```

Comment:
I'm open to suggestions! I kinda like this implementation because the `recursive=True` won't recursively search unless the user adds a **, but it allows that functionality if the user wants it. I do add a note about this functionality in the docstring so I'm not aware of anything that's overly hidden to the user.